Send binary data and small improvements.

medape opened the pull request:
First of all, I'm very sorry I have left with no warning for a year and a half. I've been meaning to write code for quite some time, but I have had a lot of things going on on my life lately. I'm fine and healthy, which is a lot these days. How have you been doing?

This pull request... possibly has the cleanest history I've ever written. Some context:
- The "hack to reliably pass binary data" is sending bytes as a string from JavaScript as usual, but forcing on them an old 8-bit encoding on both JavaScript and Java. It should be possible to simply send bytes[] but I wasn't able to make it work.
MessagePorts would probably be a cleaner way but that would require another rewrite of both sides, and you'd be forced to drop compatibility with Lollipop. I don't know if there is a backport for older APIs.
- I've added a function to fake an `alert()` from the WebView and remove some lines of code from the JavaScript. I'd actually like to move the touch pressure calibration as well; that way settings can persist through different gamepads and gamepad designers don't need to worry about that issue. But I haven't the slightest idea on how to do it.
- I've mentioned making styles more sane but the end result is a bit different. The good things about it is that Yoke errors are displayed in Yoke's blue and JavaScript prompts in white, so they're even easier to distinguish. On the other hand, it's a bit harsh on the eyes and I'd like some opinions before taking more drastic decisions.
- The app now downloads the manifest every time it connects to a webserver. It's a guarantee you're actually connecting to your webserver on your wifi once you know about it (and it reminds me to actually enable wi-fi on my phone) but it needs a better way to detect timeouts.
- There are a couple of bugs I haven't been able to properly isolate or fix yet (for example, the app usually fails to make a connection the first time you use the spinner after launching it).

Commit:
Bump Gradle version

Commit:
Add hack to reliably pass binary data

Commit:
Fix rmst/yoke-android#7

Also tweak a little the buttons on the top bar and sanitize style definitions

Commit:
Bump targetSdkVersion

Yoke runs correctly on Android 10 (actually LineageOS 17.1) even after
targetSdkVersion was raised.

Commit:
Add a function so that WebViews can emit alert()

Prompts from the Webview have a title and are colored white, so that they
can be more easily distinguished against normal Yoke prompts.

Commit:
Refactor Asynctask and "ping" webserver

To detect if the direction exists and is running Yoke, the app downloads a
lightweight file. In this way, it can display an error message if it fails.

This reduces frustration due to user error when Yoke is not running or the
wi-fi is disabled. No, I wouldn't know by experience, why do you ask.

Commit:
Change some error toasts to error popups

Commit:
Translate to Spanish Spanish

Also translated to Brazilian Portugese the new message I created.

pzmarzly commented on app/src/main/java/com/simonramstedt/yoke/YokeActivity.java:
```diff
@@ -142,19 +157,37 @@ private void deleteRecursively(File joypadPath) throws IOException {
         @JavascriptInterface
         public void update_vals(String vals) {
-            vals_str = vals;
+            vals_buffer = vals.getBytes(Charset.forName("ISO-8859-1"));
```
Why is `Charset.forName("ISO-8859-1")` preferred over `StandardCharsets.UTF_8` or `StandardCharsets.US_ASCII`?

medape replied:
Both here and in the JS side, we need a Charset that assigns a character to every byte (0x00 to 0xFF) and viceversa,

UTF-8 is a variable-length encoding. Some characters map to 2-byte sequences or more, and some bytes don't map to any characters. The result is that Python receives more bytes than expected, the `struct.unpack()` will raise an exception and the whole program crashes.

US_ASCII only maps 128 characters to the first 128 bytes (0x00 to 0x7F); as soon as we tried to decode any character outside that range, either: a) the app would crash, or b) the first bit in every byte would be lost, which would tamper with the controls.

pzmarzly replied:
Ah, I see. So, if I understand correctly, we are building joypad state as ISO-8859 garbage string, then converting it to UTF-16 on JS side via [TextDecoder](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/decode), then back to bytes on Java side. That's cool.

We are breaking compatibility with Android < 5.0 with it, though, as its native webview doesn't have TextDecoder. In my personal opinion this is okay, and we should consider bumping the required Android version to 6.0 if it'll help with development.

medape replied:
HAHAHAHAH. That's... almost correct, yes. The joypad state is built as pure binary data, and then converted to a garbage string. Which is even worse.

According to `build.gradle`, this was already incompatible with Android < 5.0 (probably because of a WebView problem). At this point, Lollipop is ancient history anyways, so... I'd actually be happy to bump the Android minimum version.

I'll try the MessagePort approach and push a couple commits to each repository if I make it.

pzmarzly commented on build.gradle:
```diff
@@ -5,7 +5,7 @@ buildscript {
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.5.0'
+        classpath 'com.android.tools.build:gradle:4.0.1'
```
Android Studio 4.1 now gives us a warning about gradle:

https://github.com/rmst/yoke-android/blob/b51bed9ceacc9238c3ca935dae290f600ed92f6f/app/src/main/java/com/simonramstedt/yoke/YokeActivity.java#L620-L621

> Resource IDs will be non-final in Android Gradle Plugin version 5.0, avoid using them in switch case statements.

I guess that means we should change this switch to if.

medape replied:
Does this help? f134e5f765644c4ed58c40b6eed334cbe15f0a42

pzmarzly replied:
Looks good to me.

pzmarzly commented on app/src/main/java/com/simonramstedt/yoke/YokeActivity.java:
```diff
@@ -183,59 +216,69 @@ protected void onPreExecute() {
+                    }
+                    return UPDATE_SUCCESS;
+                } else {
+                    // download just one file
```
This comment confused me a bit at first, as from what I understand, the file is not saved anywhere, and this conditional branch purpose is to just check if the server is online (and serves the right file)?

medape replied:
Yes. Exactly that.

Maybe in the future we could actually parse the file to compare versions or something, but for now, the file is just discarded.

pzmarzly reviewed (commented):
It's awesome to see you back in this project, and to hear you are well. As you have probably already seen, not much is being done in Yoke nowadays. So I'm really happy you are still willing to work on it, and I'll try to give you feedback in a timely manner. Of course, since you are effectively the only active developer of Yoke, and I haven't written much in either Java or Python, treat my words as just suggestions. Overall the changes look good to me, take a look at the comments I left, and once this and https://github.com/rmst/yoke/pull/56 are ready to be merged, just ping me (or just write a comment).

pzmarzly commented:
> There are a couple of bugs I haven't been able to properly isolate or fix yet (for example, the app usually fails to make a connection the first time you use the spinner after launching it).

By the way, feel free to open as many issues in this repo as you want, so that we won't forget about bugs and plans for improvements.

medape commented:
> It's awesome to see you back in this project, and to hear you are well. As you have probably already seen, not much is being done in Yoke nowadays. So I'm really happy you are still willing to work on it, and I'll try to give you feedback in a timely manner. Of course, since you are effectively the only active developer of Yoke, and I haven't written much in either Java or Python, treat my words as just suggestions.

Thank you so much. :relaxed: Yes, I checked in from time to time. I could see that the same error was reported a couple times (because the user tried to use the app on F-Droid with the latest Python code).

About that: any user can initiate a process to update an app in F-Droid (or to leave it in auto-update). After this is merged, I can try to leave things set up so that the app on F-Droid will be updated as soon as you or @rmst tag a commit as release version, if you want.

> Overall the changes look good to me, take a look at the comments I left, and once this and rmst/yoke#56 are ready to be merged, just ping me (or just write a comment).

After reading your comments, please don't merge anything yet. This week I'll try to bump the minimum required version so I can write cleaner code.

Commit:
Fix Android Studio's linter warning

Resource IDs will be non-final in Android Gradle Plugin version 5.0,
avoid using them in switch case statements

Commit:
Cleanup spinner code

and fix a bug where the spinner didn't respond because it fired extra events

medape commented:
I think we'll have to keep the garbage string for now.

I couldn't find any sane way to pass arbitrary binary data between JavaScript and Java. MessagePorts between both are always serialized to a string, and WebSockets would require installing extra libraries (and some, if not all of them, still serialize contents to a string).

Most places I checked read that you can easily pass a single primitive to Java (byte, int, double...) but you can't pass any array of them unless, again, you convert it first to a string. You could, in principle, send 2-byte or 3-byte updates and let Java reassemble them, but I doubt that really improves much the code, especially when you factor in motion controls.

The good news is I fixed almost all bugs in the spinner. As far as I checked, it only fails now when I introduce manually a direction that's already on the list.

Feel free to merge both PRs.
